### DISCUSSION TITLE
How to start with UI on the Mac?

### DISCUSSION THREAD

**Discussion by d5364t54ytfr4** [autc04/Retro68#255] [Q&A]
Like are there any libraries/headers for working with UI?

**Comment by autc04** [comment]
You will want to browse old documentation about the "Window Manager", "Menu Manager", "Dialog Manager" and "Control Manager", and maybe about the "TextEdit package".

For Retro68-specific sample code, look at Samples/Dialog, Samples/WDEF (ignoring the stuff that is actually about WDEFs) and maybe LaunchAPPL/Server; and, especially if you're using Nix, at the https://github.com/autc04/Retro68NixSample repository.

By modern standards, Apple's old UI libraries are very basic and low-level.
There were several object oriented frameworks (Apple's MacApp, Symantec's THINK Class Library, and Metrowerk's PowerPlant) built on top of it, but none of these have been ported to run on Retro68 yet, and doing so is probably not suitable as a first introduction to classic Mac GUI development ;-)

**Reply by mathieutozer** [reply]
What would it take to get PowerPlant working? From my limited understanding it is built on top of the Universal Libraries and Interfaces, so assuming you had access to that and the PowerPlant sources, could it compile?

**Reply by autc04** [reply]
I see PowerPlant has been open-sourced at the end of its life - https://github.com/chrisballinger/powerplant, so that would be a good starting point.

You'd need to spend a bit of time on build system setup, getting the include paths right (basically, throw everything in there, classic mac software didn't use relative paths in #include statements).
And then it probably *won't* compile.
But it should *almost* compile.

There might be some differences between an early c++98 implementation by Metrowerks and a recent GCC. Expect some compiler errors where you will wonder "why did it accept that code?".
It should be possible to work through this by going from error to error, I don't expect too much trouble.

There's also bound to be some problems with some central defines (PP_Macros.h) that might make wrong assumptions about what features GCC supports.

And another small detail I'd be worried about is how modern tooling deals with PowerPlant's use of `.cp` as a file extension for C++ files. That was a Mac-specific tradition that was always unheard of on other platforms.


But on the whole, it should be quite doable.

**Reply by mathieutozer** [reply]
Hey just seeing this and thanks for replying!  I got pretty far with it one evening. I was able to get a whole bunch compiling without errors, and the errors I came across were mostly easy to fix. I manually changed the cp files to cc to avoid that issue. One error that I ignored just to carry on was with `pascal` being annotated on the return types of some functions... any hints on how to resolve those? Just commenting out worked but I expect that will probably bite me once I try to get any of this code actually running. When I find some spare time I'll keep going with it and come back to update here if I have any success.

**Comment by mathieutozer** [comment]
@autc04 I am close (I think!) but am getting some linker errors for functions defined in `universal/CIncludes/fp.h`, such as `num2dec`. Any idea how to fix?

**Reply by autc04** [reply]
num2dec was something Apple-specific.
According to Google it was documented in Inside Mac: PPC Numerics.
I don't know right now where it was implemented - maybe in the MathLib library? (try -lMathLib).

**Reply by mathieutozer** [reply]
I'm a a bit of a noob when it comes to CMakeLists, how do I specify it? My usual helper (AI) isn't giving me much here. I think it needs to link to the file `universal/lib68k/libMath64Lib.a`. I tried using `target_link_libraries` but that seems to be for libraries for which I have the source code for?

**Reply by mathieutozer** [reply]
Looks like the following works

    set(MATHLIB_NAME "MathLib")
    set(MATHLIB_PATH "../../universal/lib68k/MathLib.far.o")
    add_library(${MATHLIB_NAME} STATIC ${MATHLIB_PATH})
    set_target_properties(${MATHLIB_NAME} PROPERTIES LINKER_LANGUAGE C)
    target_link_libraries(Raytracer2 ${MATHLIB_NAME})

and for reference 

`nm -a /Users/tozer/code/Retro68/universal/lib68k/MathLib.far.o` shows the list of functions in that library

**Comment by mathieutozer** [comment]
@autc04 Having another crack at linking this together. I've gathered that `MacWindows.h` is Carbon only, however I am not building a carbon app and am therefore the equivalent headers in https://github.com/chrisballinger/powerplant/blob/master/PowerPlant/Utility%20Classes/UTBAccessors.h are being compiled and used because I am setting 

```
add_definitions(-DACCESSOR_CALLS_ARE_FUNCTIONS=0)
add_definitions(-DOPAQUE_TOOLBOX_STRUCTS=0)
```

But then I still get duplicate definition errors when compiling. I had gotten the whole lot to compile by setting `DACCESSOR_CALLS_ARE_FUNCTIONS` to true, but then of course I can't link `CarbonAccessors.o` because it is PPC only, and I really want my code to run on 68k. Am I out of luck or is there some config I have wrong. The error I get is 

```
/root/Samples/Raytracer/PowerPlant/UtilityClasses/UTBAccessors.h:227:9: error: redefinition of 'Rect* GetWindowPortBounds(WindowRef, Rect*)'
In file included from /Retro68-build/toolchain/universal/CIncludes/LowMem.h:64,
                 from /root/Samples/Raytracer/PowerPlant/UtilityClasses/UTBAccessors.h:25:
/Retro68-build/toolchain/universal/CIncludes/MacWindows.h:5546:32: note: 'Rect* GetWindowPortBounds(WindowRef, Rect*)' previously defined here
 5546 |     inline DEFINE_API( Rect *) GetWindowPortBounds(WindowRef window, Rect *bounds) { return ((*(bounds) = (window)->portRect)), (bounds); }
      |                                ^~~~~~~~~~~~~~~~~~~
/root/Samples/Raytracer/PowerPlant/UtilityClasses/UTBAccessors.h:243:12: error: redefinition of 'GrafPort* GetWindowFromPort(CGrafPtr)'
/Retro68-build/toolchain/universal/CIncludes/MacWindows.h:5569:36: note: 'GrafPort* GetWindowFromPort(CGrafPtr)' previously defined here
 5569 |     inline DEFINE_API( WindowRef ) GetWindowFromPort(CGrafPtr port) { return (WindowRef)port; }
      |                                    ^~~~~~~~~~~~~~~~~
make[2]: *** [CMakeFiles/Raytracer2.dir/build.make:202: CMakeFiles/Raytracer2.dir/PowerPlant/SupportClasses/LMouseTracker.cc.obj] Error 1
make[1]: *** [CMakeFiles/Makefile2:85: CMakeFiles/Raytracer2.dir/all] Error 2
```

Which is pretty clear. Do I need to omit MacWindows.h somehow? I'll post my CMakeList.txt in case you have any insights. 

```
cmake_minimum_required(VERSION 3.22)
project(ClassicNotion)
if(APPLE)

else()    
    add_definitions(-DTARGET_CARBON=0)
    add_definitions(-DTARGET_API_MAC_OS8=1)
    add_definitions(-DTARGET_OS_MAC=1)
    add_definitions(-DACCESSOR_CALLS_ARE_FUNCTIONS=0)
    add_definitions(-DOPAQUE_TOOLBOX_STRUCTS=0)
    
    execute_process(COMMAND ls -la ../../../ OUTPUT_VARIABLE dir_contents ERROR_QUIET)
    message("Directory Contents: \n${dir_contents}")
    
    add_application(Raytracer2
        AppResources.r
        raytracer2.cc
        fixed.h
        fixed.cc
        macroman.h
        macroman.cc
        cJSON/cJSON.h
        cJSON/cJSON.c
        Block.h
        Block.cc
        PowerPlant/Headers/PP_Constants.cc
        # SupportClasses
        # PowerPlant/SupportClasses/UWMgr20Desktop.cc
        PowerPlant/SupportClasses/LClipboard.cc
        PowerPlant/SupportClasses/LException.cc
        PowerPlant/SupportClasses/LGrowZone.cc
        PowerPlant/SupportClasses/LMouseTracker.cc
        PowerPlant/SupportClasses/LRadioGroup.cc
        # PowerPlant/SupportClasses/LSIOUXAttachment.cc
        PowerPlant/SupportClasses/LString.cc
        PowerPlant/SupportClasses/LTabGroup.cc
        # PowerPlant/SupportClasses/UDesktop.cc Only include one or the other Desktop.cc files
        PowerPlant/SupportClasses/UFloatingDesktop.cc
        PowerPlant/SupportClasses/LException.cc
        # UtilityClasses
        PowerPlant/UtilityClasses/UAppearance.cc
        PowerPlant/UtilityClasses/UAttachments.cc
        # PowerPlant/UtilityClasses/UCarbonPrinting.cc
        PowerPlant/UtilityClasses/UClassicPrinting.cc
        PowerPlant/UtilityClasses/UControlMgr.cc
        PowerPlant/UtilityClasses/UCursor.cc
        PowerPlant/UtilityClasses/UDebugging.cc
        PowerPlant/UtilityClasses/UDebugging.cc
        PowerPlant/UtilityClasses/UDrawingState.cc
        PowerPlant/UtilityClasses/UDrawingUtils.cc
        PowerPlant/UtilityClasses/UEnvironment.cc
        PowerPlant/UtilityClasses/UException.cc
        PowerPlant/UtilityClasses/UException.cc
        PowerPlant/UtilityClasses/UKeyFilters.cc
        PowerPlant/UtilityClasses/UMemoryMgr.cc
        PowerPlant/UtilityClasses/UMemoryMgr.cc
        # PowerPlant/UtilityClasses/UPrinting.cc
        PowerPlant/UtilityClasses/UProcessMgr.cc
        # PowerPlant/UtilityClasses/UProfiler.cc
        PowerPlant/UtilityClasses/UReanimator.cc
        PowerPlant/UtilityClasses/URegions.cc
        PowerPlant/UtilityClasses/URegistrar.cc
        PowerPlant/UtilityClasses/UResourceMgr.cc
        PowerPlant/UtilityClasses/UScrap.cc
        PowerPlant/UtilityClasses/UScreenPort.cc
        # PowerPlant/UtilityClasses/USessionPrinting.cc
        PowerPlant/UtilityClasses/UTextTraits.cc
        PowerPlant/UtilityClasses/UWindows.cc
        # ArrayClasses
        PowerPlant/ArrayClasses/
        PowerPlant/ArrayClasses/LArray.cc
        PowerPlant/ArrayClasses/LArrayIterator.cc
        PowerPlant/ArrayClasses/LComparator.cc
        PowerPlant/ArrayClasses/LRunArray.cc
        PowerPlant/ArrayClasses/LVariableArray.cc
        # FeatureClasses
        PowerPlant/FeatureClasses/LAttachable.cc
        PowerPlant/FeatureClasses/LAttachment.cc
        PowerPlant/FeatureClasses/LBroadcaster.cc
        PowerPlant/FeatureClasses/LDragAndDrop.cc
        PowerPlant/FeatureClasses/LDragTask.cc
        PowerPlant/FeatureClasses/LEventDispatcher.cc
        PowerPlant/FeatureClasses/LListener.cc
        PowerPlant/FeatureClasses/LPeriodical.cc
        PowerPlant/FeatureClasses/LSharable.cc
        # ActionClasses
        PowerPlant/ActionClasses/UTEViewTextAction.cc
        PowerPlant/ActionClasses/LAction.cc
        PowerPlant/ActionClasses/LSimpleAEAction.cc
        PowerPlant/ActionClasses/LUndoer.cc
        PowerPlant/ActionClasses/UTETextAction.cc
        # MenuClasses
        # PowerPlant/MenuClasses/LMercutioMBAR.cc
        PowerPlant/MenuClasses/LMenu.cc
        PowerPlant/MenuClasses/LMenuBar.cc
        # FileAndStreamClasses
        PowerPlant/FileAndStreamClasses/LDataStream.cc
        PowerPlant/FileAndStreamClasses/LFile.cc
        PowerPlant/FileAndStreamClasses/LFileStream.cc
        PowerPlant/FileAndStreamClasses/LHandleStream.cc
        PowerPlant/FileAndStreamClasses/LPreferencesFile.cc
        PowerPlant/FileAndStreamClasses/LStream.cc
        # PaneClasses
        PowerPlant/PaneClasses/LActiveScroller.cc
        PowerPlant/PaneClasses/LButton.cc
        PowerPlant/PaneClasses/LCaption.cc
        PowerPlant/PaneClasses/LCicnButton.cc
        PowerPlant/PaneClasses/LControl.cc
        PowerPlant/PaneClasses/LDialogBox.cc
        PowerPlant/PaneClasses/LEditField.cc
        PowerPlant/PaneClasses/LFocusBox.cc
        PowerPlant/PaneClasses/LGrafPortView.cc
        PowerPlant/PaneClasses/LGroupBox.cc
        PowerPlant/PaneClasses/LIconPane.cc
        PowerPlant/PaneClasses/LListBox.cc
        PowerPlant/PaneClasses/LMultiPanelView.cc
        PowerPlant/PaneClasses/LOffscreenView.cc
        PowerPlant/PaneClasses/LOverlappingView.cc
        PowerPlant/PaneClasses/LPane.cc
        PowerPlant/PaneClasses/LPicture.cc
        PowerPlant/PaneClasses/LPlaceHolder.cc
        PowerPlant/PaneClasses/LPrintout.cc
        PowerPlant/PaneClasses/LRadioGroupView.cc
        PowerPlant/PaneClasses/LScroller.cc
        PowerPlant/PaneClasses/LStdControl.cc
        PowerPlant/PaneClasses/LSubOverlapView.cc
        PowerPlant/PaneClasses/LTabGroupView.cc
        PowerPlant/PaneClasses/LTable.cc
        PowerPlant/PaneClasses/LTextButton.cc
        PowerPlant/PaneClasses/LTextEditView.cc
        PowerPlant/PaneClasses/LToggleButton.cc
        PowerPlant/PaneClasses/LView.cc
        PowerPlant/PaneClasses/LWindow.cc
        PowerPlant/PaneClasses/UGWorld.cc
        PowerPlant/PaneClasses/UQuickTime.cc
        # StandardDialogs
        PowerPlant/StandardDialogs/UStandardFile.cc
        PowerPlant/StandardDialogs/LFileTypeList.cc
        PowerPlant/StandardDialogs/UClassicDialogs.cc
        PowerPlant/StandardDialogs/UConditionalDialogs.cc
        PowerPlant/StandardDialogs/UModalDialogs.cc
        PowerPlant/StandardDialogs/UNavServicesDialogs.cc
        # CommanderClasses
        PowerPlant/CommanderClasses/LSingleDoc.cc
        PowerPlant/CommanderClasses/LApplication.cc
        PowerPlant/CommanderClasses/LCommander.cc
        PowerPlant/CommanderClasses/LDocApplication.cc
        PowerPlant/CommanderClasses/LDocument.cc
        # AppleEventClasses
        PowerPlant/AppleEventClasses/UExtractFromAEDesc.cc
        PowerPlant/AppleEventClasses/LModelDirector.cc
        PowerPlant/AppleEventClasses/LModelObject.cc
        PowerPlant/AppleEventClasses/LModelProperty.cc
        PowerPlant/AppleEventClasses/LSharableModel.cc
        PowerPlant/AppleEventClasses/UAppleEventsMgr.cc
        )
    link_directories(../../../universal/lib68k)
    target_link_libraries(Raytracer2 
        libAEObjectSupportLib.a
        libInterface.a 
        libMathLib.far.a 
        libNavigation.far.a
        -Wl,--no-whole-archive
    )
        
    target_include_directories(Raytracer2 PUBLIC PowerPlant/Headers)
    target_include_directories(Raytracer2 PUBLIC PowerPlant/SupportClasses)
    target_include_directories(Raytracer2 PUBLIC PowerPlant/UtilityClasses)
    target_include_directories(Raytracer2 PUBLIC PowerPlant/ArrayClasses)
    target_include_directories(Raytracer2 PUBLIC PowerPlant/FeatureClasses)
    target_include_directories(Raytracer2 PUBLIC PowerPlant/FileAndStreamClasses)
    target_include_directories(Raytracer2 PUBLIC PowerPlant/PaneClasses)
    target_include_directories(Raytracer2 PUBLIC PowerPlant/CommanderClasses)
    target_include_directories(Raytracer2 PUBLIC PowerPlant/AppleEventClasses)
    target_include_directories(Raytracer2 PUBLIC PowerPlant/MenuClasses)
    target_include_directories(Raytracer2 PUBLIC PowerPlant/ActionClasses)
    target_include_directories(Raytracer2 PUBLIC PowerPlant/StandardDialogs)
    target_compile_options(Raytracer2 PRIVATE -fpermissive -Wno-pmf-conversions)
    set_target_properties(Raytracer2 PROPERTIES 
        LINK_FLAGS "-Wl,-gc-sections -Wl,--mac-single")    
endif()

```

**Reply by autc04** [reply]
Wierd.
First, I think you should get rid of
```
    add_definitions(-DTARGET_CARBON=0)
    add_definitions(-DTARGET_API_MAC_OS8=1)
    add_definitions(-DTARGET_OS_MAC=1)
    add_definitions(-DACCESSOR_CALLS_ARE_FUNCTIONS=0)
    add_definitions(-DOPAQUE_TOOLBOX_STRUCTS=0)
```

You're setting them to the default, but you should not need to set them at all.

Next, `MacWindows.h` is not Carbon-only. It's just that at some point, Apple renamed `Windows.h` to `MacWindows.h` and made `Windows.h` just include `MacWindows.h` internally. In short, what's in MacWindows.h is good and right and you won't be able to do without it. If there's trouble, blame it on PowerPlant, not MacWindows.h.

---

OK, so what's going on here?
PowerPlant seems to try to plaster over some differences between different target platforms (classic and carbon/mac os x) and either getting that wrong or only getting it right for some *other* version of the universal interfaces.

So with Carbon, Apple introduced some accessor functions for things that used to be just `struct` members. 
When ACCESSOR_CALLS_ARE_FUNCTIONS is 1 (set automatically by ConditionalMacros.h on Carbon), those are all declared as real functions; their implementations  are in Carbon, and in CarbonAccessors.o.
And then they defined *some* of those functions as inline functions or macros when ACCESSOR_CALLS_ARE_FUNCTIONS is 0.
Meanwhile, PowerPlant defined some of those functions in UTBAccessors. Apparently, some of Apple's inline functions got added after the last time the PowerPlant people tested compiling for classic MacOS?

My suggestion: just fix UTBAccessors.h.
To be precise, zap everything in that header file that causes a duplicate definition error.

**Reply by mathieutozer** [reply]
Thanks for the detailed response @autc04, I'm learning a lot as I go with cmake etc. 

While I have you...

Incidentally while I was waiting for a response I got https://github.com/facebook/yoga to compile without errors which could be kind of interesting to see if I can build UI with it. I did see a comment in the shared library sample though

``` 
This example does nothing useful, but it does it using a shared library.
# Doesn't run on 68K.
```

Does this mean that instead of defining the library and using it, I'd just have to lump all the source files inside my application (like I am with PowerPlant)?

I am primarily a UI engineer so am having some crazy ideas of making some kind of ComponentKit like declarative library for classic Macs, although I'm sure it would probably take about 100 years to resize a window ;-)

**Reply by autc04** [reply]
No *shared* libraries on 68K, but one could of course create and use a static library using standard CMake functionality. See `add_library` in make docs.

Mac OS 8 and 9 had a 68K port of the Code Fragment Manager to handle shared libraries, but that never caught on much because 68K was already on its way out, and anyway it'd be a big project to support in Retro68. And before that, there was the "Apple Shared Library Manager", which was strange by modern standards and which I've never seen in actual use at all.

Hmmm, yes it would be interesting to see how far it can be taken. Well, the original NeXTstep ran on 68K, and it was already more advanced than all Apple ever did before Mac OS X...

I know that some of the old hardware was physically incapable of shuffling all the pixels of a window around quickly enough to allow for really smooth dragging of a window. But it was close, scrolling a few pixels at a time worked. And of course, the Old Ones refrained from double-buffering because they wanted to use their RAM for other things besides the UI itself, and they were apparently scared of asynchronous stuff...

**Reply by mathieutozer** [reply]
Great information thanks - and of course `add_library` should suffice for my purposes.

**Reply by mathieutozer** [reply]
@autc04 - me again... I'm stuck with these linker errors. The problem is clear, AEObjectSupportLib and Interface won't link because of the symbols are more than 32kb away (16-bit relative calls). I've tried breaking my source code up into smaller static libraries and reordering, tried recompiling the toolchain with -mlong-calls but it doesn't work because the 16 bit limit is baked into the object files. I am not really sure if I have any other options at this point.
Is there some method whereby I could write some shims or something that would make the jumps for me?

```
[ 95%] Linking CXX executable Raytracer2.code.bin
/Users/tozer/code/Retro68-build/toolchain/lib/gcc/m68k-apple-macos/12.2.0/../../../../m68k-apple-macos/lib/libAEObjectSupportLib.a(AEObjectSupportLib.o): in function L335.EVALWHOSE':
(.text.L335.EVALWHOSE+0x142): relocation truncated to fit: R_68K_PC16 against symbol NEWHANDLE' defined in .text.NEWHANDLE section in /Users/tozer/code/Retro68-build/toolchain/lib/gcc/m68k-apple-macos/12.2.0/../../../../m68k-apple-macos/lib/libInterface.a(Interface.o)
/Users/tozer/code/Retro68-build/toolchain/lib/gcc/m68k-apple-macos/12.2.0/../../../../m68k-apple-macos/lib/libAEObjectSupportLib.a(AEObjectSupportLib.o): in function L327.MARKORREMEMBER':
(.text.L327.MARKORREMEMBER+0x4c): relocation truncated to fit: R_68K_PC16 against symbol GETHANDLESIZE' defined in .text.GETHANDLESIZE section in /Users/tozer/code/Retro68-build/toolchain/lib/gcc/m68k-apple-macos/12.2.0/../../../../m68k-apple-macos/lib/libInterface.a(Interface.o)
/Users/tozer/code/Retro68-build/toolchain/lib/gcc/m68k-apple-macos/12.2.0/../../../../m68k-apple-macos/lib/libAEObjectSupportLib.a(AEObjectSupportLib.o): in function L332.REDOLISTORMARK':
(.text.L332.REDOLISTORMARK+0x112): relocation truncated to fit: R_68K_PC16 against symbol SETHANDLESIZE' defined in .text.SETHANDLESIZE section in /Users/tozer/code/Retro68-build/toolchain/lib/gcc/m68k-apple-macos/12.2.0/../../../../m68k-apple-macos/lib/libInterface.a(Interface.o)
(.text.L332.REDOLISTORMARK+0x100): relocation truncated to fit: R_68K_PC16 against symbol BLOCKMOVE' defined in .text.BLOCKMOVE section in /Users/tozer/code/Retro68-build/toolchain/lib/gcc/m68k-apple-macos/12.2.0/../../../../m68k-apple-macos/lib/libInterface.a(Interface.o)
/Users/tozer/code/Retro68-build/toolchain/lib/gcc/m68k-apple-macos/12.2.0/../../../../m68k-apple-macos/lib/libAEObjectSupportLib.a(AEObjectSupportLib.o): in function L298.CREATEWHOSE':
(.text.L298.CREATEWHOSE+0x100): relocation truncated to fit: R_68K_PC16 against symbol BLOCKMOVE' defined in .text.BLOCKMOVE section in /Users/tozer/code/Retro68-build/toolchain/lib/gcc/m68k-apple-macos/12.2.0/../../../../m68k-apple-macos/lib/libInterface.a(Interface.o)
/Users/tozer/code/Retro68-build/toolchain/lib/gcc/m68k-apple-macos/12.2.0/../../../../m68k-apple-macos/lib/libAEObjectSupportLib.a(AEObjectSupportLib.o): in function L278.CLEARNEWHANDLE':
(.text.L278.CLEARNEWHANDLE+0x12): relocation truncated to fit: R_68K_PC16 against symbol NEWHANDLECLEAR' defined in .text.NEWHANDLECLEAR section in /Users/tozer/code/Retro68-build/toolchain/lib/gcc/m68k-apple-macos/12.2.0/../../../../m68k-apple-macos/lib/libInterface.a(Interface.o)
/Users/tozer/code/Retro68-build/toolchain/lib/gcc/m68k-apple-macos/12.2.0/../../../../m68k-apple-macos/lib/libAEObjectSupportLib.a(AEObjectSupportLib.o): in function L340.ISVALIDHASHTABLE':
(.text.L340.ISVALIDHASHTABLE+0x10): relocation truncated to fit: R_68K_PC16 against symbol `GETHANDLESIZE' defined in .text.GETHANDLESIZE section in /Users/tozer/code/Retro68-build/toolchain/lib/gcc/m68k-apple-macos/12.2.0/../../../../m68k-apple-macos/lib/libInterface.a(Interface.o)
collect2: error: ld returned 1 exit status
make[2]: *** [Raytracer2.code.bin] Error 1
make[1]: *** [CMakeFiles/Raytracer2.dir/all] Error 2
make: *** [all] Error 2
```

**Reply by mathieutozer** [reply]
Ok little update - I worked out that PP only needs `AEObjectSupportLib` and `Interface` to support Apple Events (Scripting) which I do not care about, so I took an axe to the source and removed that dependency and it linked! 

My next step is to get some resources on screen. I have a menu bar appear, and I was able to affect the Quit item by just plonking `exit(0)` instead of the `SendAEQuit();`  which makes me confident that the PP environment is somewhat running.

Back in March I was able to convert a .rsrc file to .r using some tool on modern macOS but I forget which, care to remind me?
--- Found it `DeRez`

PP also uses some ppob resource files which I think are proprietary project files for their constructor app, but I think it ends up converting them to plain .r files, so might be nice to figure out how to do that as well.

**Comment by superdave** [comment]
On Apr 14, 2025, at 2:45 PM, Wolfgang Thaller ***@***.***> wrote:
> 
> 
> No shared libraries on 68K, but one could of course create and use a static library using standard CMake functionality. See add_library in make docs.
> 
> Mac OS 8 and 9 had a 68K port of the Code Fragment Manager to handle shared libraries, but that never caught on much because 68K was already on its way out, and anyway it'd be a big project to support in Retro68. And before that, there was the "Apple Shared Library Manager", which was strange by modern standards and which I've never seen in actual use at all.

ISTR some versions of MS Office used it?

> Hmmm, yes it would be interesting to see how far it can be taken. Well, the original NeXTstep ran on 68K, and it was already more advanced than all Apple ever did before Mac OS X...

Sure, but NeXTstep also took proper advantage of the virtual memory capabilities of the 68K, which MacOS never did (it would have broken too much about how applications expected to work by the time it was feasible hardware-wise), and it's difficult to do shared libraries efficiently if you can't just map them in like Unix and Windows do.

> I know that some of the old hardware was physically incapable of shuffling all the pixels of a window around quickly enough to allow for really smooth dragging of a window. But it was close, scrolling a few pixels at a time worked. And of course, the Old Ones refrained from double-buffering because they wanted to use their RAM for other things besides the UI itself, and they were apparently scared of asynchronous stuff...

I mean, yeah, but also double-buffering takes up enormous CPU resources when you don't have hardware page-flipping (which isn't practical for windowed double-buffering anyway).  There is, however, plenty of asynchronous stuff in the old Toolbox, it's just done via callbacks.  Drawing procedures for windows and controls and filter procedures are really great examples.

Event handling was primarily done as a loop because honestly that's how it's done a lot of places; you can always abstract that to be asynchronous via your own callbacks if you want, but handling it as a queue consumer is actually a lot more efficient from an overhead standpoint.
